File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* search(TreeNode* root, TreeNode* p, TreeNode* q){
      if(root == NULL) return NULL;
    if(p->val < root->val && q->val < root->val) return search(root->left,p,q);
    if(p->val > root->val && q->val > root->val) return search(root->right,p,q);
    return root;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
      
        return search(root,p,q);
        
    }
};

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
   TreeNode ans = null;
    void find(TreeNode root, TreeNode p, TreeNode q){
       if(root == null) return;
       if(root.val == p.val || root.val == q.val && ans == null){
        ans = root;
        return;
       }
     
       if(p.val < root.val && q.val < root.val){
         find(root.left, p,q);
       }
       else if(q.val > root.val && p.val > root.val){
          find(root.right, p,q);
       }
       else if(ans == null){
        ans = root;
        return;

       }
     
      

    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null) return null;
        if(root.left == null && root.right == null) return root;
        find(root, p , q);
        return ans;
    }
}